pkg/controller/disruption/disruption.go

```diff
@@ -0,0 +1,587 @@
+	// GetPodReplicaSets returns an error only if no ReplicaSets are found.  We
+	// don't return that as an error to the caller.
+	if err != nil {
+		return cas, nil
+	}
+	controllerScale := map[types.UID]int32{}
+	for _, rs := range rss {
+		// GetDeploymentsForReplicaSet returns an error only if no matching
+		// deployments are found.  We don't return that as an error to the caller.
+		ds, err := dc.dLister.GetDeploymentsForReplicaSet(&rs)
+		if err != nil {
+			return cas, nil
+		}
+		for _, d := range ds {
+			controllerScale[d.UID] = d.Spec.Replicas
+		}
+	}
+
+	for uid, scale := range controllerScale {
+		cas = append(cas, controllerAndScale{UID: uid, scale: scale})
+	}
+
+	return cas, nil
+}
+
+func (dc *DisruptionController) getPodReplicationControllers(pod *api.Pod) ([]controllerAndScale, error) {
+	cas := []controllerAndScale{}
+	rcs, err := dc.rcLister.GetPodControllers(pod)
+	if err == nil {
+		for _, rc := range rcs {
+			cas = append(cas, controllerAndScale{UID: rc.UID, scale: rc.Spec.Replicas})
+		}
+	}
+	return cas, nil
+}
+
+func (dc *DisruptionController) Run(stopCh <-chan struct{}) {
+	glog.Infof("Starting disruption controller")
+	if dc.kubeClient != nil {
+		glog.V(0).Infof("Sending events to api server.")
+		dc.broadcaster.StartRecordingToSink(dc.kubeClient.Events(""))
+	} else {
+		glog.V(0).Infof("No api server defined - no events will be sent to API server.")
+	}
+	go dc.pdbController.Run(stopCh)
+	go dc.podController.Run(stopCh)
+	go dc.rcController.Run(stopCh)
+	go dc.rsController.Run(stopCh)
+	go dc.dController.Run(stopCh)
+	go wait.Until(dc.worker, time.Second, stopCh)
+	<-stopCh
+	glog.Infof("Shutting down disruption controller")
+}
+
+func (dc *DisruptionController) addDb(obj interface{}) {
+	pdb := obj.(*policy.PodDisruptionBudget)
+	glog.Infof("add DB %q", pdb.Name)
+	dc.enqueuePdb(pdb)
+}
+
+func (dc *DisruptionController) updateDb(old, cur interface{}) {
+	pdb := cur.(*policy.PodDisruptionBudget)
+	glog.Infof("update DB %q", pdb.Name)
+	dc.enqueuePdb(pdb)
+}
+
+func (dc *DisruptionController) removeDb(obj interface{}) {
+	pdb := obj.(*policy.PodDisruptionBudget)
+	glog.Infof("remove DB %q", pdb.Name)
+	dc.enqueuePdb(pdb)
+}
+
+func (dc *DisruptionController) addPod(obj interface{}) {
+	pod := obj.(*api.Pod)
+	glog.Infof("addPod called on pod %q", pod.Name)
+	pdb := dc.getPdbForPod(pod)
+	if pdb == nil {
+		glog.Infof("No matching pdb for pod %q", pod.Name)
+		return
+	}
+	glog.Infof("addPod %q -> PDB %q", pod.Name, pdb.Name)
+	dc.enqueuePdb(pdb)
+}
+
+func (dc *DisruptionController) updatePod(old, cur interface{}) {
+	pod := cur.(*api.Pod)
+	glog.Infof("updatePod called on pod %q", pod.Name)
+	pdb := dc.getPdbForPod(pod)
+	if pdb == nil {
+		glog.Infof("No matching pdb for pod %q", pod.Name)
+		return
+	}
+	glog.Infof("updatePod %q -> PDB %q", pod.Name, pdb.Name)
+	dc.enqueuePdb(pdb)
+}
+
+func (dc *DisruptionController) deletePod(obj interface{}) {
+	pod := obj.(*api.Pod)
```

Comment:
This may be a `DeleteFinalStateUnknown` so you need to do something like [this](https://github.com/kubernetes/kubernetes/blob/afe4977c9badb90f8c883d76fac07ab5c26d13d9/pkg/controller/deployment/deployment_controller.go#L382)